### DISCUSSION TITLE
How to display pictures correctly in ItemsRepeater

### DISCUSSION THREAD

**Discussion by howzew** [AvaloniaUI/Avalonia#17613] [Q&A] [answered]
I tried to show some pictures in the ItemsRepeater and here's my code:

```
<ScrollViewer>
	<ItemsRepeater ItemsSource="{Binding ImageList}">
		<ItemsRepeater.Layout>
			<UniformGridLayout ItemsStretch="Uniform" />
		</ItemsRepeater.Layout>
	        <ItemsRepeater.ItemTemplate>
	                <DataTemplate>
		                 <CheckBox >
			                    <Image Source="{Binding Image}" Stretch="UniformToFill" Width="100" Height="100"/>
		                   </CheckBox>
                          </DataTemplate>
	           </ItemsRepeater.ItemTemplate>
	</ItemsRepeater>
</ScrollViewer>
```

I have two pictures, a square and a rectangle, and they don't show up correctly at the same time

When I set the image to Stretch="UniformToFill", both images are stretched and cannot be displayed in their entirety.

![uniformtofill](https://github.com/user-attachments/assets/d57bc5bd-3b51-45ee-9664-d6a0abeaa8be)

When I set the image to Stretch="Fill", the square ones will display normally, and the rectangular ones will be squeezed.

![fill](https://github.com/user-attachments/assets/01ec1b70-a091-439a-856f-855b7370c784)

When I set the image to Stretch="Uniform", the rectangular ones will display normally, while the square ones will be squeezed.

![uniform](https://github.com/user-attachments/assets/9bf8080b-c790-4c54-a4db-5a5841c1d0cb)

How can we make pictures of different shapes scale normally in the container without being stretched or squeezed?

**Comment by maxkatz6** [comment]
Uniform:
![image](https://github.com/user-attachments/assets/45096afd-a9a5-413c-9e18-a8db9cfc6ad6)

**Reply by maxkatz6** [reply]
You should double check if your square image doesn't have any extra paddings.

**Reply by howzew** [reply]
@maxkatz6  My code is above and the picture does not have any extra paddings

**Reply by thevortexcloud** [reply]
Does it work if you remove the fixed width/height?

**Reply by howzew** [reply]
@thevortexcloud The image will be stretched in some way I don't know the rules, and it will still be stretched.

**Reply by thevortexcloud** [reply]
Hmm strange. I always rescaled things myself and used a fixed width since it used less RAM and sidestepped a lot of weird virtualisation problems with items repeater. I posted the code I used here:

https://github.com/AvaloniaUI/Avalonia/discussions/13605#discussioncomment-10794788

**Reply by howzew** [reply]
I have seen this code before and tried it, but it still doesn't solve my problem

**Reply by howzew** [reply]
@maxkatz6  I think this is a bug, at least a bug in uniform, because uniform should at least keep the ratio of width to height, which is also true from the example image you gave, but in my example, image "G", obviously does not keep the correct aspect ratio

**Reply by thevortexcloud** [reply]
It's possible. If it is it's unlikely to be fixed, at least officially. Although a PR may still be accepted for it. The code is just a port from UWP IIRC. So there may also be a hint in the UWP docs.

https://learn.microsoft.com/en-us/windows/apps/design/controls/items-repeater

That said I am not a member of the Avalonia team, so they may say something else.

**Reply by howzew** [reply]
It may not be a bug of ItemsRepeater, because I tried using Uniform in ListBox and ItemsControl, and the image also cannot keep the aspect ratio.

**Comment by thevortexcloud** [comment]
On a side note: ItemsRepeater is being deprecated soonish.

https://github.com/AvaloniaUI/Avalonia.Controls.ItemsRepeater

**Comment by howzew** [comment]
@timunie Why doesn't the ‘Uniform ’ keep the correct width and height ratio in my example?

**Reply by timunie** [reply]
Can you please provide the original pictures?

**Reply by howzew** [reply]
![G1](https://github.com/user-attachments/assets/29dc9f0a-ad06-4130-bf5b-544137f6b90a)
![G2](https://github.com/user-attachments/assets/5b8dada6-68aa-4b44-8932-77f95ad28c50)
yes,and I tried using Uniform in ListBox and ItemsControl, and the image also cannot keep the aspect ratio.

**Reply by timunie** [reply]
Github doesn't let me download the image. :-(

**Reply by timunie** [reply]
![image](https://github.com/user-attachments/assets/fdecf980-c94f-4f5c-bd39-2dd192e8bd5b)

Working perfectly for me 🤷

**Reply by howzew** [reply, marked as answer]
I'm very sorry, I finally found the reason. It's because I specified the resolution when generating my image, which caused this problem. Thank you.
